File: graph_and_trees/amount_of_time_for_binary_tree_to_be_infected.cpp
//
// Created by Harsha Vardhan on 23/08/22.
//

#include <iostream>
#include <unordered_set>
#include <unordered_map>
#include <queue>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

void constructParentMap(TreeNode *root, unordered_map<TreeNode *, TreeNode *> &parentMap, int &start, TreeNode *&s) {
    if (root->val == start) {
        s = root;
    }
    if (root->left) {
        parentMap[root->left] = root;
        constructParentMap(root->left, parentMap, start, s);
    }
    if (root->right) {
        parentMap[root->right] = root;
        constructParentMap(root->right, parentMap, start, s);
    }
}

int amountOfTime(TreeNode *root, int start) {
    TreeNode *l = root;
    unordered_map<TreeNode *, TreeNode *> parentMap;
    parentMap[root] = nullptr;
    TreeNode *s = nullptr;
    constructParentMap(l, parentMap, start, s);

    queue<TreeNode *> q;
    unordered_set<int> visited;

    q.push(s);
    visited.insert(start);

    int count = 0;

    while (!q.empty()) {
        int size = q.size();
        count++;

        for (int i = 0; i < size; i++) {
            TreeNode *front = q.front();
            q.pop();

            if (front->left && visited.find(front->left->val) == visited.end()) {
                visited.insert(front->left->val);
                q.push(front->left);
            }

            if (front->right && visited.find(front->right->val) == visited.end()) {
                visited.insert(front->right->val);
                q.push(front->right);
            }

            TreeNode *t = parentMap[front];
            if (t && visited.find(t->val) == visited.end()) {
                visited.insert(t->val);
                q.push(t);
            }
        }
    }
    return count - 1;
}

int main() {
    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(5);
    root->right = new TreeNode(3);
    root->left->right = new TreeNode(4);
    root->left->right->left = new TreeNode(9);
    root->left->right->right = new TreeNode(2);
    root->right->left = new TreeNode(10);
    root->right->right = new TreeNode(6);

    cout << amountOfTime(root, 3) << "\n";
    return 0;
}

File: go/tree/amount_of_time_for_binary_tree_to_be_infected.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

//type Queue[T any] struct {
//	data []T
//}
//
//func (q *Queue[T]) Empty() bool {
//	return q.Len() == 0
//}
//
//func (q *Queue[T]) Len() int {
//	return len(q.data)
//}
//
//func (q *Queue[T]) Push(inp T) {
//	q.data = append(q.data, inp)
//}
//
//func (q *Queue[T]) Pop() T {
//	if q.Len() == 0 {
//		panic("queue is empty")
//	}
//	e := q.data[0]
//	q.data = q.data[1:]
//	return e
//}
//
//func (q *Queue[T]) Front() T {
//	if q.Len() == 0 {
//		panic("queue is empty")
//	}
//	return q.data[0]
//}

func dfs(root *TreeNode, parentMap map[int]*TreeNode, startNode *TreeNode, start int) {
	if root.Val == start {
		*startNode = *root
	}

	if root.Left != nil {
		parentMap[root.Left.Val] = root
		dfs(root.Left, parentMap, startNode, start)
	}

	if root.Right != nil {
		parentMap[root.Right.Val] = root
		dfs(root.Right, parentMap, startNode, start)
	}
}

func bfs(root *TreeNode, parentMap map[int]*TreeNode) int {
	q := Queue[*TreeNode]{}
	q.Push(root)
	depth := 0
	visited := make(map[int]struct{})
	visited[root.Val] = struct{}{}

	for !q.Empty() {
		size := q.Len()

		for i := 0; i < size; i++ {
			t := q.Pop()
			if t.Left != nil {
				if _, ok := visited[t.Left.Val]; !ok {
					q.Push(t.Left)
					visited[t.Left.Val] = struct{}{}
				}
			}
			if t.Right != nil {
				if _, ok := visited[t.Right.Val]; !ok {
					q.Push(t.Right)
					visited[t.Right.Val] = struct{}{}
				}
			}
			parent := parentMap[t.Val]
			if parent != nil {
				if _, ok := visited[parent.Val]; !ok {
					q.Push(parent)
					visited[parent.Val] = struct{}{}
				}
			}
		}

		depth++
	}
	return depth - 1
}

func amountOfTime(root *TreeNode, start int) int {
	if root == nil {
		return 0
	}
	parentMap := make(map[int]*TreeNode)
	parentMap[root.Val] = nil

	startNode := &TreeNode{}
	dfs(root, parentMap, startNode, start)

	return bfs(startNode, parentMap)
}

func main() {
	var head *TreeNode = &TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val: 5,
			Right: &TreeNode{
				Val:   4,
				Left:  &TreeNode{Val: 9},
				Right: &TreeNode{Val: 2},
			},
		},
		Right: &TreeNode{
			Val:   3,
			Left:  &TreeNode{Val: 10},
			Right: &TreeNode{Val: 6},
		},
	}
	fmt.Println(amountOfTime(head, 3))
}
